keps/sig-auth/2579-psp-replacement/README.md

```diff
@@ -0,0 +1,1039 @@
+publish the following tools:
+
+- Library implementation - The admission plugin will be written using reusable library code.
+- Webhook implementation - A standalone webhook implementation will be provided (using the same
+  library code), enabling policy enforcement on older clusters.
+- Thorough testing resources, which can be used to validate 3rd party implementations of the policy
+  spec.
+
+### Test Plan
+
+The admission controller can safely be enabled as a no-op with the default-defaults, i.e. everything
+is privileged. This will let us run the admission controller in our standard E2E test jobs, by
+relabeling specific test namespaces.
+
+**E2E Tests:** The following tests should be added:
+
+1. Allow mode tests:
+    - Test all profile levels
+    - Test profile version support
+2. Warning mode tests:
+    - Profile levels & version support
+3. Namespace policy relabeling
+    - Ensure labeling completes even when there are warnings
+    - Test warning on violating pods
+    - Test dry-run mode
+
+Additionally, we should add tests to the upgrade test suite to ensure that version skew is properly
+handled:
+
+- A minimally specified pod (just a container image) should always be allowed by the baseline
+  policy.
+- A privileged pod should never be allowed by baseline or restricted
+- A Fully specified pod within the bounds of baseline should be allowed by baseline, and rejected by
+  restricted.
+- A minimally specified restricted pod should be allowed at a pinned version.
+
+**Integration Tests:** Audit mode tests should be added to integration testing, where we have
+existing audit logging tests.
+
+**Manual Testing Resources:** Pod resources will be provided covering all dimensions of the baseline
+& restricted profiles, for validation of 3rd party policy implementations. These have been drafted
+by @JimBugwadia: https://github.com/JimBugwadia/pod-security-tests
+
+**Unit Tests:** Both the library and admission controller implementations will have thorough
+coverage of unit tests.
+
+### Monitoring
+
+<<[UNRESOLVED]>>
+
+**BLOCKING**
+
+The following metrics could be useful:
+
+1. Policy decision - record the policy decision for each mode, for each pod create/update request.
+   Potential label dimensions:
+    - decision (allow, deny, exempt)
+    - policy mode (enforcing, warning, audit)
+    - policy level (privileged, baseline, restricted)
+    - policy version (latest, current, past, future), optionally replace current & past with the
+      specific version
+    - resource type (Pod, Deployment, ReplicaSet, etc.) - for audit & warnings applied to templated
+      pods
+    - request type (create, update)
+2. Exemptions - record whenever a request is exempt (should this only increment when it would
+   otherwise have been denied?)
+    - exempt dimension (user, namespace, runtimeclass)
+3. (maybe) Violated control - record the specific policy check that failed.
```

Comment:
How would you monitor this?  If i'm reading it correctly you want to find violations that weren't caught by the admission controller?

Reply:
No, this would be violations that _are_ caught by admission control. In other words, this is a finer-grained view of metric 1 above in the case of deny, i.e. why was the request denied?

This could be useful if a user wanted to set up an alert for specific types of denials (e.g. pod requesting priviliged access). There is a lot of overlap with the audit log here though, so I'm on the fence about whether we should include metrics for this, or just point users to the audit log for more detailed information.

Reply:
I think this is too fine-grained for metrics... same for resource type... you'll still need to hit the audit log to figure out the particular resources that failed to resolve the issue anyway

Reply:
Yeah, I agree for the specific control - use the audit logs instead. For resource type, I could see it being useful at pod/not-pod granularity, but since templated pods are only checked for warning & audit levels, maybe it doesn't matter...